sentiment_analysis/pytorch/train.py

```diff
@@ -0,0 +1,179 @@
+from torchtext import datasets
+import torch
+import tqdm
+import torch.nn as nn
+from torch.autograd import Variable
+import torch.nn.functional as F
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+def IMDB_dataset(use_cuda=True, batch_size=128, max_len=2470):
+    """
+    Pytorch generator for IMDB dataset.
+    Args:
+    use_cuda - bool
+    batch_size - int
+    max_len - int - max length of the sentence in train.
+    All smaller sentences will be padded to have length = max_len.
+    All larger sentences will be cropped.
+    Returns:
+    train_iter, test_iter - batch generators
+    len(TEXT.vocab) - vocabulary size. Necessary for the embedding layer
+    batch_size
+    """
+    device = "cuda:0" if use_cuda else -1
+    # set up fields
+    TEXT = data.Field(lower=True, fix_length=max_len, batch_first=True)
+    LABEL = data.Field(sequential=False)
+    # make splits for data
+    train, test = datasets.IMDB.splits(TEXT, LABEL)
+    # build the vocabulary
+    TEXT.build_vocab(train)
+    LABEL.build_vocab(train)
+    train_iter, test_iter = data.BucketIterator.splits(
+    (train, test), batch_size=batch_size, device=device)
+    return train_iter, test_iter, len(TEXT.vocab), batch_size
+
+
+class ConvNet(nn.Module):
+    """
+    Architecture:
+    Embedding layer with customizable vocab size and
+    embeding size,
+    2d Convolutional layer, 
+    2d Max Pooling layer,
+    Fully-connected layer
+    """
+    def __init__(self, vocab_size, embedding_size=1024):
+        super().__init__()
+        self.vocab_size = vocab_size
+        self.embedding_size = embedding_size
+        self.features = nn.Sequential(
+            nn.Conv2d(1, 1, kernel_size=3),
+            nn.ELU(),
+            nn.MaxPool2d(kernel_size=3),
+        )
+        self.classifier = nn.Linear(279480, 2)
+        
+    def forward(self, sentences):
+        embed_layer = nn.Embedding(self.vocab_size, self.embedding_size, sparse=True)
+        sentences_embedded = embed_layer(sentences)
+        sentences_embedded = sentences_embedded.unsqueeze(1)
+        out = self.features(sentences_embedded)
+        out = out.view(out.size(0), -1)
+        out = self.classifier(out)
+        return F.log_softmax(out, dim=-1)
+
+
+def train(model, optimizer, n_epochs, train_iter, test_iter, vocab_size, batch_size, quality):
+    """
+    Performs training with n_epochs steps.
+    """
+    train_log, train_acc_log = [], []
+    val_log, val_acc_log = [], []
+
+    for epoch in range(n_epochs):
+        train_loss, train_acc = train_epoch(model, optimizer, train_iter)
+        
+        val_loss, val_acc = test(model, test_iter, quality)
+        
+        train_log.extend(train_loss)
+        train_acc_log.extend(train_acc)
+
+        steps = 25000 / batch_size
+        val_log.append((steps * (epoch + 1), np.mean(val_loss)))
+        val_acc_log.append((steps * (epoch + 1), np.mean(val_acc)))
+        
+        plot_history(train_log, val_log)
```

Comment:
plotting will be a performance hit to the training, i don't think it is desirable in benchmark... correct me if i'm wrong.